DatadogRUM/Sources/Integrations/TelemetryReceiver.swift

```diff
@@ -211,12 +258,19 @@ internal final class TelemetryReceiver: FeatureMessageReceiver {
             if sessionId != self.currentSessionID {
                 self.currentSessionID = sessionId
                 self.eventIDs = []
+                self.eventsCount = 0
             }
 
-            // record up to `maxEventsPerSessions`, discard duplicates
-            if self.eventIDs.count < TelemetryReceiver.maxEventsPerSessions, !self.eventIDs.contains(id) {
-                self.eventIDs.insert(id)
-                operation(context, writer)
+            // record up to `maxEventsPerSessions`, discard duplicates for events with `id`
+            if self.eventsCount < TelemetryReceiver.maxEventsPerSessions {
+                if id == nil {
+                    self.eventsCount += 1
+                    operation(context, writer)
+                } else if let eventID = id, !self.eventIDs.contains(eventID) {
+                    self.eventIDs.insert(eventID)
+                    self.eventsCount += 1
+                    operation(context, writer)
+                }
```

Comment:
```suggestion
                if let eventID = id, !self.eventIDs.contains(eventID) {
                    self.eventIDs.insert(eventID)
                }
                self.eventsCount += 1
                operation(context, writer)
```

Reply:
It isn't the same. Note that the original code allows one empty branch: when `id != nil`, but it is already stored in `self.eventIDs` it should do nothing (no increment and no `operation()`).

The suggested code doesn't pass `testSendTelemetry_discardDuplicates()` test, which was a good exercise 🙂👍.

Reply:
💯